python/tvm/autotvm/graph_tuner/base_graph_tuner.py

```diff
@@ -0,0 +1,508 @@
+                    flops *= i
+                inferred_time = flops * avg_time
+                record_input = MeasureInput(target=self._target, task=None, config=None)
+                record_output = MeasureResult(costs=(inferred_time,), error_no=0,
+                                              all_cost=-1, timestamp=-1)
+                self._layout_transform_dict[ltf_workload] = (record_input, record_output)
+                continue
+
+            records = []
+            task = autotvm.task.create(layout_transform, args=args, target=self._target,
+                                       target_host=target_host)
+            task.workload = ltf_workload
+            tuner = autotvm.tuner.GridSearchTuner(task)
+            tuner.tune(n_trial=1, measure_option=measure_option,
+                       callbacks=[_log_to_list(records)])
+            if not isinstance(records[0][1].costs[0], float):
+                records[0] = (records[0][0], records[0][1]._replace(costs=(INVALID_LAYOUT_TIME,)))
+            self._layout_transform_dict[ltf_workload] = records[0]
+
+        self._iterate_layout_transform(self._create_matrix_callback)
+        self._logger.info("Benchmarking layout transformation successful.")
+
+    @property
+    def layout_transform_dict(self):
+        """Get layout transformation dictionary for input graph.
+
+        Returns
+        -------
+        layout_transform_dict : dict of tuple to (MeasureInput, MeasureResult)
+            Layout transformation dictionary for input graph.
+        """
+        return self._layout_transform_dict
+
+
+    def get_optimal_records(self):
+        """Convert optimal record dictionary to a list of records
+        with ascending order of node index in graph.
+
+        Returns
+        -------
+        sch_list : list of tuple
+            List of records with ascending order of node index in graph.
+        """
+        ordered_index_list = sorted(self._optimal_record_dict.keys())
+        ret = []
+        for index in ordered_index_list:
+            node_entry = self._node_list[index]
+            if node_entry["op"] not in self._target_ops:
+                continue
+            ret.append(node_entry["record_candidates"][self._optimal_record_dict[index]])
+        return ret
+
+    def write_opt_sch2record_file(self, record_file="graph_opt_schedule.log"):
+        """Write graph level optimal schedules into file.
+
+        Parameters
+        ----------
+        record_file : str, optional
+            Output schedule file.
+        """
+        with open(record_file, "a") as out_file:
+            records = self.get_optimal_records()
+            for record in records:
+                out_file.write(encode(record[0], record[1]) + "\n")
+        msg = "Writing optimal schedules to %s successfully." % record_file
+        self._logger.info(msg)
+
+    @abstractmethod
+    def run(self, **kwargs):
+        """Run graph tuning."""
+        pass
```

Comment:
better to throw exception?

Reply:
Since this is defined as abstractmethod, we don't need another exception throwing?